dhis2/dhis2-core
# feat: execute jobs as specific user [DHIS2-15166]

### Summary
Adds a persisted  `executedBy` property to `JobConfiguration`.
It **can** be set to the user that should be running the `Job`.
If set, a `SecurityContext` for that user is wrapped around the `Job#execute` call.
No form of password check or any other actual authentication is performed to obtain the user's `SecurityContext` .
Instead 
* system jobs run without an user as superuser
* user created jobs can set any specific user or none to run as superuser
* programmatic created jobs can set any specific user or none to run as superuser

### Automatic Testing
There are tests using the `executeNow` method of the scheduling manager. 
In tests this always has done a synchronous `execute` on the job. 
Since this now will set and clear the `SecurityContext` the context present before the call needs to be restored afterwards as this is not actually running in a worker thread as it would in production.

### Manual Testing
Triggering data integrity manually will run the data integrity checks asynchronously using a job as the user that triggered it.
* `POST` to `/api/dataIntegrity?checks=categories_unique_category_combo` with user auth header (either no `Content-Type` header or `application/json` as value even for empty body)
* check `/api/scheduling/completed/DATA_INTEGRITY` - the root object in the array now has a `userId` property referring to the user that triggered the job (this ID is extracted from the security context so it means execution occurs as that user)

In contrast when running data integrity as a persisted job by creating a configuration in the scheduler app and running it manually the same result of that run `/api/scheduling/completed/DATA_INTEGRITY` does not have a `userId` as the job runs as superuser.
